src/sentry/types/region.py

```diff
@@ -119,42 +120,145 @@ class RegionContextError(Exception):
+            if not regions_with_monolith_name:
+                raise RegionConfigurationError(
+                    self._error_msg_for_region_setting(
+                        "SENTRY_MONOLITH_REGION", monolith_region_name
+                    )
+                )
+            (self.monolith_region,) = regions_with_monolith_name
+
+        self._by_name = {r.name: r for r in self.regions}
+
+        self.local_region: Region | None = None
+        if SiloMode.get_current_mode() == SiloMode.REGION:
+            if not settings.SENTRY_REGION:
+                raise RegionConfigurationError(
+                    "SENTRY_REGION must be set when server is in REGION silo mode"
+                )
+            try:
+                self.local_region = self._by_name[settings.SENTRY_REGION]
+            except KeyError:
+                raise RegionConfigurationError(
+                    self._error_msg_for_region_setting("SENTRY_REGION", settings.SENTRY_REGION)
+                )
+
+    def _error_msg_for_region_setting(self, name: str, value: str) -> str:
+        return (
+            f"The {name} setting (value={value!r}) must point to a region name "
+            f"(region names = {[r.name for r in self.regions]!r})"
+        )
 
-        self.regions = frozenset(regions)
-        self.by_name = {r.name: r for r in self.regions}
+    def get(self, region_name: str) -> Region | None:
+        return self._by_name.get(region_name)
 
     def validate_all(self) -> None:
         for region in self.regions:
             region.validate()
 
 
-def parse_raw_config(region_config: Any) -> Iterable[Region]:
+class GlobalRegionDirectory:
+    """The set of all regions in this Sentry platform instance.
+
+    This class is a singleton. In a production environment, it is immutable.
+    Test environments support swapping its values in a temporary context.
+    """
+
+    def __init__(self, directory: RegionDirectory) -> None:
+        self._dir = directory
+        self._temporary_local_region: Region | None = None  # Used only in test env
+
+    @property
+    def regions(self) -> frozenset[Region]:
+        return self._dir.regions
+
+    @property
+    def historic_monolith_region(self) -> Region:
+        return self._dir.monolith_region
+
+    @property
+    def local_region(self) -> Region | None:
+        if self._temporary_local_region is not None:
+            self._allow_only_in_test_env()
+            return self._temporary_local_region
+        return self._dir.local_region
+
+    def get(self, region_name: str) -> Region | None:
+        return self._dir.get(region_name)
+
+    def get_all_region_names(self) -> Iterable[str]:
+        return (r.name for r in self.regions)
+
+    def _allow_only_in_test_env(self) -> None:
+        if not in_test_environment():
+            raise Exception("Swapping region values is allowed only in the test environment")
```

Comment:
Weekend shower thought to pick up on Monday: the intent here is to ensure that these methods are called only from `testutils`. But if we did break that rule, unit tests would nonetheless pass (a false negative), because calling from a production context would still count as `in_test_environment()`. The check is still better than nothing, as we'd get a rapid signal in the form of prod errors so as to provoke a rollback, but obviously that's not really what we want.

Perhaps a better approach would be to move the "swap" methods from this class to a subclass that exists only in `testutils`. Rely on, I don't know, maybe a fixture to swap in the subclass. Then functions like `assume_test_silo_mode` could assume/assert that it has an instance of the subclass and call the methods appropriately. Someone who erroneously tries to call the test-only methods outside of `testutils` would still be getting the instance that has those methods, so we still get a false negative at runtime for the same reasons. But at least Mypy would complain about it.

Another, even jankier option would be to inspect the call stack and check that `testutils` is in there somewhere. Then, merely being in a unit test wouldn't be enough to cause a false negative (we'd see `tests` but not necessarily `testutils`). At least that's true some of the time, when `testutils` doesn't happen to be there for some other reason.

Reply:
> Perhaps a better approach would be to move the "swap" methods from this class to a subclass that exists only in testutils

Having the swapping logic live in `testutils` is nice because we already have existing tests that ensure production code doesn't import symbols from `testutils`.